corehq/apps/analytics/static/analytics/js/logging.js

```diff
@@ -0,0 +1,176 @@
+ * VERBOSE: ANALYTICS_VERBOSE
+ * WARNING, DEPRECATED: ANALYTICS_DEPRECATED
+ *
+ */
+hqDefine('analytics/js/logging', function () {
+    'use strict';
+    var _LEVELS = {
+        warning: ['WARNING', 'color: #994f00;'],
+        verbose: ['VERBOSE', 'color: #685c53;'],
+        debug: ['DEBUG', 'color: #004ebc;'],
+        deprecated: ['DEPRECATED', 'color: #e53e30;'],
+    };
+
+    var _printPretty = function (message) {
+        var _title, _value, _grp;
+        if (_.isUndefined(message.value)) {
+            console.log("Message was undefined");  // eslint-disable-line no-console
+        } else if (message.value.isGroup) {
+            _grp = message.value;
+            _grp.level = message.level;
+            _grp.print();
+        } else if (_.isArguments(message.value)) {
+            _title = "Arguments (" + message.value.length + ")    " + _.map(Array.from(message.value), JSON.stringify).join('    ');
+            _value = Array.from(message.value);
+            _grp = new Group(_title, new Message(_value, message.level));
+            _grp.isCollapsed = true;
+            _grp.print();
+        } else if (_.isArray(message.value)) {
+            _.each(message.value, function (msg) {
+                _printPretty(new Message(msg, message.level));
+            });
+        } else if (_.isObject(message.value) && _.has(message.value, 0) && _.isElement(message.value[0])) {
+            // DOM element
+            _title = "#" + message.value.get(0).id + " ." + message.value.get(0).className.split(' ').join('.');
+            _value = message.value.get(0).outerHTML;
+            _grp = new Group(_title, new Message(_value, message.level));
+            _grp.isCollapsed = true;
+            _grp.print();
+        } else if (!_.isString(message.value) && !_.isUndefined(message.value)) {
+            _title = JSON.stringify(message.value);
+            _value = message.value;
+            _grp = new Group(_title, new Message(_value, message.level));
+            _grp.isCollapsed = true;
+            _grp.isRaw = true;
+            _grp.print();
+        } else {
+            message.print();
+        }
+    };
+
+    var _getStyle = function (level) {
+        var levelOptions = _LEVELS[level];
+        return (levelOptions) ? levelOptions[1] : "";
+    };
+
+    var Message = function (value, level) {
+        var msg = {};
+        msg.level = level;
+        msg.value = value;
+        msg.print = function () {
+            console.log(msg.value);  // eslint-disable-line no-console
+        };
+        return msg;
+    };
+
+    var Group = function (title, message) {
+        var grp = {};
+        grp.title = title;
+        grp.level = message.level;
+        grp.message = message;
+        grp.isCollapsed = false;
+        grp.isRaw = false;
+        grp.isGroup = true;
+        grp.print = function () {
+            var _printGrp = (grp.isCollapsed) ? console.groupCollapsed : console.group;  // eslint-disable-line no-console
```

Comment:
This is cool, I didn't know `console.group[Collapsed]` existed.

Reply:
me too! I also didn't know you could use CSS to style `console.log` statements!! mind blown 💥